src/interpreter/plugin/AddressPlugin.ts

```diff
@@ -0,0 +1,53 @@
+/**
+ * @license
+ * Copyright (c) 2023 Handsoncode. All rights reserved.
+ */
+
+import {columnIndexToLabel} from '../../parser/addressRepresentationConverters'
+import {ProcedureAst} from '../../parser'
+import {InterpreterState} from '../InterpreterState'
+import {InterpreterValue} from '../InterpreterValue'
+import {FunctionArgumentType, FunctionPlugin, FunctionPluginTypecheck, ImplementedFunctions} from './FunctionPlugin'
+
+enum AbsStyle {
+  FullyAbsolute = 1,
+  RowAbsoluteColRelative = 2,
+  RowRelativeColAbsolute = 3,
+  FullyRelative = 4,
+}
+
+export class AddressPlugin extends FunctionPlugin implements FunctionPluginTypecheck<AddressPlugin> {
+  public static implementedFunctions: ImplementedFunctions = {
+    'ADDRESS': {
+      method: 'address',
+      parameters: [
+        {argumentType: FunctionArgumentType.NUMBER, minValue: 1},
+        {argumentType: FunctionArgumentType.NUMBER, minValue: 1},

```

Comment:
We should allow negative values for these arguments because they are valid in R1C1 references. For `=ADDRESS(-1, -1, 4, FALSE)` Excel returns `R[-1]C[-1]`.

Also, please add unit tests for calling `ADDRESS` with negative col and row numbers. For both R1C1 and A1 notation styles and for different absolute/relative modes.

Reply:
Oh - nice catch.

I had used the `FunctionPlugin` argument checking facility in this implementation because the [OpenDoc  spec pre-conditioned](http://docs.oasis-open.org/office/v1.2/os/OpenDocument-v1.2-os-part2.html#__RefHeading__1018416_715980110)  `row` and `column` arguments as being `>=1` - but you are correct to note that that Excel and Sheets allow values `<1` (**sometimes**)

**Frustratingly** - it appears that range checking used by both Excel and Sheets is variable depending on the other argument values: `row` and `column` must always be `>=1` if `A1` style is requested; but when requesting `R1C1` style the range checking depend on the requested `MODE`.

| ROW      | COL      | MODE    | STYLE     | ADDRESS   |
| -------- | ------- | --------- | --------- | ------------ | 
|-1| 	-1| 	1| 	FALSE| 	#VALUE!  |
|-1| 	-1| 	2| 	FALSE| 	#VALUE! |
| -1| 	-1| 	3| 	FALSE| 	#VALUE!  |
| -1	| -1| 	4| 	FALSE| 	R[-1]C[-1] |

| ROW      | COL      | MODE    | STYLE     | ADDRESS   |
| -------- | ------- | --------- | --------- | ------------ | 
| -1| 	1	| 1| 	FALSE| 	#VALUE! |
| -1	| 1| 	2	| FALSE| 	#VALUE! |
| -1	| 1| 	3| 	FALSE| 	R[-1]C1 |
| -1	| 1| 	4| 	FALSE| 	R[-1]C[1] |

| ROW      | COL      | MODE    | STYLE     | ADDRESS   |
| -------- | ------- | --------- | --------- | ------------ | 
| 1| 	-1| 	1| 	FALSE| 	#VALUE! |
| 1| 	-1| 	2| 	FALSE| 	R1C[-1] |
| 1| 	-1| 	3| 	FALSE| 	#VALUE! |
| 1| 	-1| 	4| 	FALSE| 	R[1]C[-1] |

So it looks like the arg checking provided by `FunctionPlugin` won't work and I'll need to roll custom argument validation based on the other argument values. 

Will work on that now - and I will add a big table test that exhaustively checks all argument combinations.

Thanks for catching this!

Reply:
Added support for negative (and zero) row and col indicators; Added full test matrix against Excel and Sheets